crates/flux-fhir-analysis/src/wf/sortck.rs

```diff
@@ -336,30 +361,40 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
                 self.check_expr(e2, &sort)?;
                 Ok(rty::Sort::Bool)
             }
-            fhir::BinOp::Mod => {
-                self.check_expr(e1, &rty::Sort::Int)?;
-                self.check_expr(e2, &rty::Sort::Int)?;
-                Ok(rty::Sort::Int)
-            }
+
             fhir::BinOp::Lt | fhir::BinOp::Le | fhir::BinOp::Gt | fhir::BinOp::Ge => {
                 let sort = self.next_sort_var();
                 self.check_expr(e1, &sort)?;
                 self.check_expr(e2, &sort)?;
 
-                // Elaborate sort of operator
-                let sort = self
-                    .fully_resolve(&sort)
-                    .map_err(|_| self.emit_err(errors::CannotInferSort::new(expr.span)))?;
+                // Elaborate sort of operator (force as `Int` for `tests/pos/surface/forall01.rs`)
+                let sort = self.resolve_or_unify(&sort, rty::Sort::Int);
+
                 self.wfckresults
                     .bin_rel_sorts_mut()
                     .insert(expr.fhir_id, sort);
 
                 Ok(rty::Sort::Bool)
             }
-            fhir::BinOp::Add | fhir::BinOp::Sub | fhir::BinOp::Mul | fhir::BinOp::Div => {
+            fhir::BinOp::Add
+            | fhir::BinOp::Sub
+            | fhir::BinOp::Mul
+            | fhir::BinOp::Div
+            | fhir::BinOp::Mod => {
                 let sort = self.next_num_var();
                 self.check_expr(e1, &sort)?;
                 self.check_expr(e2, &sort)?;
+
+                // Elaborate sort of operator
+                let sort = self.resolve_or_unify(&sort, rty::Sort::Int);
+
+                self.wfckresults
+                    .bin_rel_sorts_mut()
+                    .insert(expr.fhir_id, sort.clone());
+
+                // check that the sort is integral for mod (and div?)
+                self.check_integral(op, &sort);
```

Comment:
hmmm `mod` is a pesky one. Z3 seems to allow `mod` between reals, but I'm unsure how that is interpreted. It also doesn't look like standard smtlib. 

I think we should require the sort to be fully resolved (locally) for `mod` (so we don't have to carry extra constraints) but allow all the other operations to proceed with possible unresolved sorts